### ISSUE TITLE
New indexes created for datastreams after update to `8.15.0` are without lifecycle policies

### ISSUE THREAD

**Issue by lahsivjar** [elastic/apm-server#13898]
**APM Server version** (`apm-server version`): `8.15.0`

**Description of the problem including expected versus actual behavior**: In `8.15.0` we migrated from ILM to [DLM](https://www.elastic.co/guide/en/elasticsearch/reference/current/data-stream-lifecycle.html). New indexes created for clusters which migrate to `8.15.0` don't have any lifecycle attached as existing datastream needs to be updated explicitly: https://www.elastic.co/guide/en/elasticsearch/reference/current/tutorial-manage-existing-data-stream.html.

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem,
including server configuration, agent(s) used, etc.  The easier you make it
for us to reproduce it, the more likely that somebody will take the time to
look at it.

 1. Create a cluster with version < `8.15.0`
 2. Send data to the cluster (for example: traces) (and continue sending it throughout the steps)
 3. Perform a rollover on the datastream getting data
 4. Upgrade the cluster to `8.15.0`
 5. Observe that the latest indices are `Unmanaged` (for example: by using `GET /_data_stream/traces-apm-default` for traces datastream)
 6. Perform a rollover on the datastream getting data
 7. Observe that the newly created index is also `Unmanaged` (for example: by using `GET /_data_stream/traces-apm-default` for traces datastream)

Temporary mitigation is to explicitly set the lifecycle using the PUT API. For example, the below operations would set data retention based on APM defaults for all APM datastreams:

```
PUT _data_stream/traces-apm-*/_lifecycle
{
  "data_retention": "10d" 
}

PUT _data_stream/traces-apm.rum*/_lifecycle
{
  "data_retention": "90d" 
}

PUT _data_stream/traces-apm.sampled*/_lifecycle
{
  "data_retention": "1h" 
}

PUT _data_stream/metrics-apm.*.1m-*/_lifecycle
{
  "data_retention": "90d" 
}

PUT _data_stream/metrics-apm.*.10m-*/_lifecycle
{
  "data_retention": "180d" 
}

PUT _data_stream/metrics-apm.*.60m-*/_lifecycle
{
  "data_retention": "390d" 
}

PUT _data_stream/metrics-apm.internal-*/_lifecycle
{
  "data_retention": "90d" 
}

PUT _data_stream/metrics-apm.app.*/_lifecycle
{
  "data_retention": "90d" 
}

PUT _data_stream/logs-apm.*/_lifecycle
{
  "data_retention": "10d" 
}
```

**Provide logs (if relevant)**: N/A

**Comment by lahsivjar** [comment]
The fix for `v8.15.1` has been merged via https://github.com/elastic/elasticsearch/pull/112097, however, the same patch cannot be applied to `v8.16.0`. The details of this are tracked in https://github.com/elastic/elasticsearch/issues/112137

**Comment by axw** [comment]
I don't love https://github.com/elastic/elasticsearch/pull/112097 as a solution, since it kinda implies we'll keep the old ILM config around in the index templates for perpetuity.

Perhaps we can remove that while also addressing serverless, by dynamically injecting the ILM config into the index template if there are existing indices with ILM config?

**Comment by axw** [comment]
We need to handle the following scenarios:

- Create a new 8.15.1+ cluster, with defaults: DLM should be used
- Create a new 8.15.1+ cluster, with customised ILM (https://www.elastic.co/guide/en/observability/current/apm-ilm-how-to.html#apm-data-streams-custom-three): customised ILM policy should be used
- Upgrade 8.14.x to 8.15.1+ with defaults: ILM should continue to be used for old indices, DLM should be used for new indices
- Upgrade 8.14.x to 8.15.1+ with customised ILM (same guide as above): ILM should continue to be used for old indices, and also for new indices

**Comment by lahsivjar** [comment]
> Create a new 8.15.1+ cluster, with defaults: DLM should be used

Has been tested in ... (link to the PR where testing was done)

> Create a new 8.15.1+ cluster, with customised ILM (https://www.elastic.co/guide/en/observability/current/apm-ilm-how-to.html#apm-data-streams-custom-three): customised ILM policy should be used

WIP

> Upgrade 8.14.x to 8.15.1+ with defaults: ILM should continue to be used for old indices, DLM should be used for new indices

WIP

> Upgrade 8.14.x to 8.15.1+ with customised ILM (same guide as above): ILM should continue to be used for old indices, and also for new indices

WIP

> (**adding another one here**) Upgrade 8.14.x to 8.15.0 with custom ILM policy: custom ILM policy should work as expected (no bugs for this case)

<detail>
<summary>Tested ✅ (the cluster has been upgraded from `8.14.3` to `8.15.0`)</summary>

```
{
  "data_streams": [
    {
      "name": "traces-apm-default",
      "timestamp_field": {
        "name": "@timestamp"
      },
      "indices": [
        {
          "index_name": ".ds-traces-apm-default-2024.09.02-000001",
          "index_uuid": "DenckzoZRdq7EricBWuz3g",
          "prefer_ilm": true,
          "ilm_policy": "traces-apm.traces-default_policy",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.02-000002",
          "index_uuid": "OnoamUPZQT6OaXGLsQEvDA",
          "prefer_ilm": true,
          "ilm_policy": "traces-apm.traces-default_policy",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.02-000003",
          "index_uuid": "tIUDNwthS3Wf2GOnigD1Rg",
          "prefer_ilm": true,
          "ilm_policy": "custom-traces-apm-policy",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.02-000004",
          "index_uuid": "p2LkzfgJSf6FQjBg2LouvQ",
          "prefer_ilm": true,
          "ilm_policy": "custom-traces-apm-policy",
          "managed_by": "Index Lifecycle Management"
        }
      ],
      "generation": 4,
      "_meta": {
        "package": {
          "name": "apm"
        },
        "managed_by": "fleet",
        "managed": true
      },
      "status": "YELLOW",
      "template": "traces-apm@template",
      "ilm_policy": "custom-traces-apm-policy",
      "next_generation_managed_by": "Index Lifecycle Management",
      "prefer_ilm": true,
      "hidden": false,
      "system": false,
      "allow_custom_routing": false,
      "replicated": false,
      "rollover_on_write": false
    }
  ]
}
```

</detail>

**Comment by simitt** [comment]
Reading through the test cases, can you clarify if this is true: customers on `8.14.3` with custom ILM policies, who upgrade to `8.15.1` will not have to do any manual interactions for their custom ILM policies to continue to be applied?

Please also provide test scenarios where users already have upgraded to `8.15.0` and then upgrade to `8.15.1`.

**Comment by lahsivjar** [comment]
> Reading through the test cases, can you clarify if this is true: customers on 8.14.3 with custom ILM policies, who upgrade to 8.15.1 will not have to do any manual interactions for their custom ILM policies to continue to be applied?

True, no changes need to be done for this case.

> Please also provide test scenarios where users already have upgraded to 8.15.0 and then upgrade to 8.15.1.

For customers having custom ILM, all would be good. Even for version 8.15.0, they would be in the clear i.e. all their indices would be managed by the configured custom ILM policy (already tested above).

However, if a customer with the default ILM policy has moved to 8.15.0 and then upgraded to 8.15.1, then, the indices created in 8.15.0 would be unmanaged even after the upgrade to 8.15.1. Indices created after the upgrade to 8.15.1 would be good though. I was initially thinking of suggesting configuring DSL as when DSL on a datastream is configured then all unmanaged indices are moved to be managed by DSL, however, this would have the side-effect of moving from ILM to DSL. I don't think this should be a big deal as the solution would only be required for installations using the default ILM policies, but, would be good if others could validate this (CC:@silvia, @axw)

**Comment by axw** [comment]
> This, as expected, creates unmanaged indices. We would need to suggest workaround for this in our changelog/release-notes. The simplest way would be to have users with default ILM settings explicitly configure DSL for all datastreams. Since default ILM and DSL are identical, there would be no impact to the users. In future, if the user wants to move to ILM, they can do so by creating a custom component template. WDYT @axw ?

It's the simplest code change, but I'm not convinced that anyone will go ahead and actively switch their data streams to DSL; I expect we'll end up with users indefinitely sticking with the legacy ILM policies and bifurcating the Serverless & Hosted experience. IMO we should take one of two routes:

 1. Automatically switch data streams with default ILM policy to DSL on upgrade. This should be non-breaking, since the retention periods match, and there's no use of hot/warm/cold in the default ILM policies.
 2. Automatically switch all data streams to DSL on upgrade. This would be breaking when users have customised ILM, so we would need to document how users can re-enable ILM by setting `prefer_ilm: true` in their `@custom` component templates.

I prefer (2) since users will need to start setting `prefer_ilm: true` in their `@custom` component templates if they want to use ILM. The only other alternative I can see is to revert the use of DSL, but I feel like the cat's already out of the bag with 8.15.0 being released.

**Comment by simitt** [comment]
>>Upgrade 8.14.x to 8.15.0 with default ILM policy and then upgrade to 8.15.1

>This, as expected, creates unmanaged indices. We would need to suggest workaround for this in our changelog/release-notes. The simplest way would be to have users with default ILM settings explicitly configure DSL for all datastreams. Since default ILM and DSL are identical, there would be no impact to the users. In future, if the user wants to move to ILM, they can do so by creating a custom component template. WDYT @axw ?

@lahsivjar it only creates unmanaged indices in `8.15.0`, right? With the fix and changes in `8.15.1`, my understanding is that as soon as users upgrade to this version, new, managed backing indices will be created. The `8.15.0` ones would still be unmanaged, but any new ones would not. Can you confirm this?

**Comment by lahsivjar** [comment]
> @lahsivjar it only creates unmanaged indices in 8.15.0, right? With the fix and changes in 8.15.1, my understanding is that as soon as users upgrade to this version, new, managed backing indices will be created. The 8.15.0 ones would still be unmanaged, but any new ones would not. Can you confirm this?

Yes, this is correct. Only indices created with version 8.15.0 will remain unmanaged.

**Comment by lahsivjar** [comment]
### Testing with BC

> Create a new 8.15.1+ cluster, with defaults: DLM should be used

<details>
<summary> ✅ Tested </summary>

```json
{
  "data_streams": [
    {
      "name": "traces-apm-default",
      "timestamp_field": {
        "name": "@timestamp"
      },
      "indices": [
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000001",
          "index_uuid": "T5YO3keARTqvNSbSOKFuJQ",
          "prefer_ilm": false,
          "ilm_policy": "traces-apm.traces-default_policy",
          "managed_by": "Data stream lifecycle"
        }
      ],
      "generation": 1,
      "_meta": {
        "managed": true,
        "description": "Index template for traces-apm-*"
      },
      "status": "YELLOW",
      "template": "traces-apm@template",
      "lifecycle": {
        "enabled": true,
        "data_retention": "10d"
      },
      "ilm_policy": "traces-apm.traces-default_policy",
      "next_generation_managed_by": "Data stream lifecycle",
      "prefer_ilm": false,
      "hidden": false,
      "system": false,
      "allow_custom_routing": false,
      "replicated": false,
      "rollover_on_write": false
    }
  ]
}
```
</details>

> Create a new 8.15.1+ cluster, with customised ILM (https://www.elastic.co/guide/en/observability/current/apm-ilm-how-to.html#apm-data-streams-custom-three): customised ILM policy should be used

<details>
<summary>✅ Tested (requires to set `"prefer_ilm": true` in the `@custom` component template)</summary>

```json
{
  "data_streams": [
    {
      "name": "traces-apm-default",
      "timestamp_field": {
        "name": "@timestamp"
      },
      "indices": [
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000001",
          "index_uuid": "T5YO3keARTqvNSbSOKFuJQ",
          "prefer_ilm": false,
          "ilm_policy": "traces-apm.traces-default_policy",
          "managed_by": "Data stream lifecycle"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000002",
          "index_uuid": "2gyoEUu7TvSdcEQm6NDJUw",
          "prefer_ilm": false,
          "ilm_policy": "traces-apm.traces-default_policy",
          "managed_by": "Data stream lifecycle"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000003",
          "index_uuid": "qRE3mbVORzKdhJ7Iy9TafQ",
          "prefer_ilm": true,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        }
      ],
      "generation": 3,
      "_meta": {
        "managed": true,
        "description": "Index template for traces-apm-*"
      },
      "status": "YELLOW",
      "template": "traces-apm@template",
      "lifecycle": {
        "enabled": true,
        "data_retention": "10d"
      },
      "ilm_policy": "custom-ilmdlm",
      "next_generation_managed_by": "Index Lifecycle Management",
      "prefer_ilm": true,
      "hidden": false,
      "system": false,
      "allow_custom_routing": false,
      "replicated": false,
      "rollover_on_write": false
    }
  ]
}
```
</details>

> Upgrade 8.14.x to 8.15.1+ with defaults: ILM should continue to be used for old indices, DLM should be used for new indices

<details>
<summary>⚠️ Tested (Note that new indices would continue to use ILM on upgrade)</summary>

```json
{
  "data_streams": [
    {
      "name": "traces-apm-default",
      "timestamp_field": {
        "name": "@timestamp"
      },
      "indices": [
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000001",
          "index_uuid": "Y4BYJa12S965zieMmkKP2g",
          "prefer_ilm": true,
          "ilm_policy": "traces-apm.traces-default_policy",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000002",
          "index_uuid": "ksh_hUWmSVCZpBNPdlePvQ",
          "prefer_ilm": true,
          "ilm_policy": "traces-apm.traces-default_policy",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000003",
          "index_uuid": "G7ujwqE8QiCY-SWxwlSitQ",
          "prefer_ilm": false,
          "ilm_policy": "traces-apm.traces-default_policy",
          "managed_by": "Index Lifecycle Management"
        }
      ],
      "generation": 3,
      "_meta": {
        "package": {
          "name": "apm"
        },
        "managed_by": "fleet",
        "managed": true
      },
      "status": "YELLOW",
      "template": "traces-apm@template",
      "ilm_policy": "traces-apm.traces-default_policy",
      "next_generation_managed_by": "Index Lifecycle Management",
      "prefer_ilm": false,
      "hidden": false,
      "system": false,
      "allow_custom_routing": false,
      "replicated": false,
      "rollover_on_write": false
    }
  ]
}
```
</details>

> Upgrade 8.14.x to 8.15.1+ with customised ILM (same guide as above): ILM should continue to be used for old indices, and also for new indices

<details>
<summary>✅ Tested</summary>

```json
{
  "data_streams": [
    {
      "name": "traces-apm-default",
      "timestamp_field": {
        "name": "@timestamp"
      },
      "indices": [
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000001",
          "index_uuid": "kU3XXLL0Riq603vV8rOHwA",
          "prefer_ilm": true,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000002",
          "index_uuid": "FPO8lBu4T8W29TcsxfAZ8g",
          "prefer_ilm": true,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000003",
          "index_uuid": "OWSm1CIVTBK8txzi4E23_g",
          "prefer_ilm": false,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        }
      ],
      "generation": 3,
      "_meta": {
        "package": {
          "name": "apm"
        },
        "managed_by": "fleet",
        "managed": true
      },
      "status": "YELLOW",
      "template": "traces-apm@template",
      "ilm_policy": "custom-ilmdlm",
      "next_generation_managed_by": "Index Lifecycle Management",
      "prefer_ilm": false,
      "hidden": false,
      "system": false,
      "allow_custom_routing": false,
      "replicated": false,
      "rollover_on_write": false
    }
  ]
}
```
</details>

> Upgrade 8.14.x to 8.15.0 with custom ILM policy: custom ILM policy should work as expected (no bugs for this case)

<details>
<summary>✅ Tested</summary>

```json
{
  "data_streams": [
    {
      "name": "traces-apm-default",
      "timestamp_field": {
        "name": "@timestamp"
      },
      "indices": [
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000001",
          "index_uuid": "FF5029kASPiBFiXQIIpHFQ",
          "prefer_ilm": true,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000002",
          "index_uuid": "hSYm1-G1QOeqc2DmEyTVGw",
          "prefer_ilm": true,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000003",
          "index_uuid": "HURNF7zhTYG_8NQGSRTPRw",
          "prefer_ilm": true,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        }
      ],
      "generation": 3,
      "_meta": {
        "package": {
          "name": "apm"
        },
        "managed_by": "fleet",
        "managed": true
      },
      "status": "YELLOW",
      "template": "traces-apm@template",
      "ilm_policy": "custom-ilmdlm",
      "next_generation_managed_by": "Index Lifecycle Management",
      "prefer_ilm": true,
      "hidden": false,
      "system": false,
      "allow_custom_routing": false,
      "replicated": false,
      "rollover_on_write": false
    }
  ]
}
```
</details>

<details>
<summary>✅ Tested (after upgrading the same cluster from 8.15.0 to 8.15.1, continues to use ILM)</summary>

```json
{
  "data_streams": [
    {
      "name": "traces-apm-default",
      "timestamp_field": {
        "name": "@timestamp"
      },
      "indices": [
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000001",
          "index_uuid": "FF5029kASPiBFiXQIIpHFQ",
          "prefer_ilm": true,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000002",
          "index_uuid": "hSYm1-G1QOeqc2DmEyTVGw",
          "prefer_ilm": true,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000003",
          "index_uuid": "HURNF7zhTYG_8NQGSRTPRw",
          "prefer_ilm": true,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000004",
          "index_uuid": "D7YUM1PdSsSpH-taPns1vw",
          "prefer_ilm": false,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000005",
          "index_uuid": "gbSyICRJQl-x0T6Dceq1XA",
          "prefer_ilm": false,
          "ilm_policy": "custom-ilmdlm",
          "managed_by": "Index Lifecycle Management"
        }
      ],
      "generation": 5,
      "_meta": {
        "package": {
          "name": "apm"
        },
        "managed_by": "fleet",
        "managed": true
      },
      "status": "YELLOW",
      "template": "traces-apm@template",
      "ilm_policy": "custom-ilmdlm",
      "next_generation_managed_by": "Index Lifecycle Management",
      "prefer_ilm": false,
      "hidden": false,
      "system": false,
      "allow_custom_routing": false,
      "replicated": false,
      "rollover_on_write": false
    }
  ]
}
```
</details>

> Upgrade 8.14.x to 8.15.0 with default ILM policy and then upgrade to 8.15.1

<details>
<summary>⚠️ Tested (indices created in 8.15.0 remain unmanaged but new ones are managed by ILM)</summary>

```json
{
  "data_streams": [
    {
      "name": "traces-apm-default",
      "timestamp_field": {
        "name": "@timestamp"
      },
      "indices": [
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000001",
          "index_uuid": "HdvHJejIRM6BepeJpoZP9Q",
          "prefer_ilm": true,
          "ilm_policy": "traces-apm.traces-default_policy",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000002",
          "index_uuid": "oxcuDgYiQu6uEzwcbjRc0w",
          "prefer_ilm": true,
          "ilm_policy": "traces-apm.traces-default_policy",
          "managed_by": "Index Lifecycle Management"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000003",
          "index_uuid": "VIhBWb41RWyObWU06CDwDQ",
          "prefer_ilm": true,
          "managed_by": "Unmanaged"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000004",
          "index_uuid": "WH4nZ_zwQySm5nqyg7fOYQ",
          "prefer_ilm": true,
          "managed_by": "Unmanaged"
        },
        {
          "index_name": ".ds-traces-apm-default-2024.09.03-000005",
          "index_uuid": "IboaSTnBTtmgYpnCLS-hPw",
          "prefer_ilm": false,
          "ilm_policy": "traces-apm.traces-default_policy",
          "managed_by": "Index Lifecycle Management"
        }
      ],
      "generation": 5,
      "_meta": {
        "package": {
          "name": "apm"
        },
        "managed_by": "fleet",
        "managed": true
      },
      "status": "YELLOW",
      "template": "traces-apm@template",
      "ilm_policy": "traces-apm.traces-default_policy",
      "next_generation_managed_by": "Index Lifecycle Management",
      "prefer_ilm": false,
      "hidden": false,
      "system": false,
      "allow_custom_routing": false,
      "replicated": false,
      "rollover_on_write": false
    }
  ]
}
```
</details>
